icicle/tests/test_hash_api.cpp

```diff
@@ -313,38 +316,210 @@ class HashSumBackend : public HashBackend
+  }
+
+  // int nof_hashes = nof_inputs / hashes[0]->m_total_input_limbs;
+  int side_inputs_offset = 0;
+  for (int i = 0; i < hashes.size(); i++) {
+    nof_outputs = nof_input_byes * hashes[i].output_size() / hashes[i].input_default_chunk_size();
+    const int nof_hashes = nof_input_byes / hashes[i].input_default_chunk_size();
+
+    auto config = default_hash_config();
+    config.batch = nof_hashes;
+    hashes[i].hash(layer_in, hashes[i].input_default_chunk_size(), config, layer_out);
+
+    // Transfer outputs to inputs before moving to the next layer
+    for (int i = 0; i < nof_outputs; i++) {
+      layer_in[i] = layer_out[i];
+    }
+    nof_input_byes = nof_outputs;
+  }
+
+  // Compare computed root with the tree's root
+  auto [root, root_size] = tree.get_merkle_root();
+
+  for (int i = 0; i < root_size; i++) {
+    ICICLE_ASSERT(root[i] == layer_out[i]) << "Mismatch in root[" << i << "]";
+  }
+
+  delete[] layer_in;
+  delete[] layer_out;
+}
+
+template <typename in_limb>
+void assert_valid_tree(
+  const MerkleTree& tree,
+  int nof_inputs,
+  const in_limb* inputs,
+  const std::vector<Hash>& hashes,
+  const MerkleTreeConfig& config)
+{
+  return assert_valid_tree(
+    tree, nof_inputs * sizeof(in_limb), reinterpret_cast<const std::byte*>(inputs), hashes, config);
+}
+
+TEST_F(HashApiTest, MerkleTreeBasic)
 {
-  ICICLE_CHECK(icicle_set_device(s_reference_target)); // TODO CUDA too
   // define input
-  constexpr int nof_leaves = 100;
-  uint32_t leaves[nof_leaves];
-  const size_t input_size = sizeof(leaves);
-  for (int i = 0; i < nof_leaves; ++i) {
-    leaves[i] = i;
+  int leaf_size = sizeof(merkel_hash_t);
+  constexpr int nof_leaves = 50;
+  merkel_hash_t leaves[nof_leaves];
+  randomize(leaves, nof_leaves);
+
+  merkel_hash_t other_leaves[nof_leaves];
+  randomize(other_leaves, nof_leaves);
+
+  // define the merkle tree
+  auto config = default_merkle_tree_config();
+  auto layer0_hash = HashSumBackend::create(5 * leaf_size, 2 * leaf_size); // in 5 leaves, out 2 leaves  200B ->  80B
+  auto layer1_hash = HashSumBackend::create(4 * leaf_size, leaf_size);     // in 4 leaves, out 1 leaf    80B  ->  20B
+  auto layer2_hash = HashSumBackend::create(leaf_size, leaf_size);         // in 1 leaf, out 1 leaf      20B  ->  20B
+  auto layer3_hash = HashSumBackend::create(5 * leaf_size, leaf_size); // in 5 leaves, out 1 leaf    20B  ->  4B output
+
+  std::vector<Hash> hashes = {layer0_hash, layer1_hash, layer2_hash, layer3_hash};
+
+  int output_store_min_layer;
+  randomize(&output_store_min_layer, 1);
+  output_store_min_layer = output_store_min_layer & 3; // Ensure index is in a valid 0-3 range
+  ICICLE_LOG_DEBUG << "Min store layer:\t" << output_store_min_layer;
+
+  auto prover_tree = MerkleTree::create(hashes, sizeof(merkel_hash_t), output_store_min_layer);
+
+  auto verifier_tree = MerkleTree::create(hashes, sizeof(merkel_hash_t), output_store_min_layer);
+
+  // build tree
+
+  // START_TIMER(MerkleTree_build)
```

Comment:
you cann keep the timer. Just remove the std::cout since the END_TIMER is already printing that